### ISSUE TITLE
Step into library import statement gives "Could not load source '<string>': Source unavailable."

### ISSUE THREAD

**Issue by luabud** [Microsoft/ptvsd#1291]
Set a breakpoint to a ``import numpy`` statement and step into it. It'll raise: ``Could not load source '<string>': Source unavailable.``

VS Code Insiders 1.33.0
Extension version 2019.3.6215

launch.json file:

```
{
    // Use IntelliSense to learn about possible attributes.
    // Hover to view descriptions of existing attributes.
    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
    "version": "0.2.0",
    "configurations": [
        
        {
            "name": "Python-test",
            "type": "python",
            "request": "launch",
            "program": "${file}",
            "console": "integratedTerminal",
            "stopOnEntry": true,
            "cwd": "${workspaceFolder}",
        },
    ]
}
```

**Comment by DonJayamanne** [comment]
You are debugging non-user code, hence you need to turn that feature on.
Add the following into `launch.json`
```js
"debugStdLib": true
```

**Comment by karthiknadig** [comment]
It should not send back '<string>' though. If filename is '<string>', we should return sourceReference. If there is no source reference, we should not return those frames.

**Comment by DonJayamanne** [comment]
> It should not send back '' though. If filename is '', we should return sourceReference. If there is no source reference, we should not return those frames.

Aah yes, agreed, I thought `step in` would just automatically perform a `step over`.
Isn't that the expected behavior. I thought thats what it did in the past.

**Comment by karthiknadig** [comment]
You are correct. `step-in` effectively becomes `step-over`, when `debugStdLib=true`. But it seems to also send `<string>` as frame. I need to investigate that.

**Comment by karthiknadig** [comment]
![image](https://user-images.githubusercontent.com/3840081/55375051-80543600-54bf-11e9-8909-be77471f2274.png)

This is what we get. I think message should describe why users are seeing this and what users can do about it. Message should be something along the lines of:
* If you intended to debug a site-package or standard library please set `debugStdLib: true` in `launch.json` for VS Code. For Visual Studio 2017/2019 set the debug option at Tools > Options > Python > Debugging.
* If you stepped into a `@dataclass` operator or `exec`, we may not be able to find the source for it. If you feel this is a bug please file a issue `https://github.com/Microsoft/ptvsd/issues` with repro steps.

**Comment by int19h** [comment]
`<string>` is not treated as stdlib, so when an stdlib module does an eval or exec, and you step in, you end up in something like this. You can see the relevant bits here: https://github.com/Microsoft/ptvsd/pull/1245/commits/d41e3a58e244d760c9044980ef06af485cf754af (note that the behavior wrt `<string>` is the same in both old and new code), and some discussion about this here: https://github.com/Microsoft/ptvsd/issues/1232#issuecomment-473075614.

We can't just exclude `<string>`, because anything that's run with `python -c` has that as the name, and then of course user code can also do eval or exec. If I remember correctly, in ptvsd 3, when it saw a frame like that, it walked one step back to see if the *previous* frame is user code or stdlib.

**Comment by karthiknadig** [comment]
Ahh. I see. So the only reason `debugStdLib: true`  feels like it works because user has not yet reached the point where the exec or eval is called.

**Comment by int19h** [comment]
Is the behavior actually different depending on `debugStdLib` setting here? I would expect it to be the same, since this one frame is not considered stdlib either way. The only difference should be that if you have stdlib on, then a step-in should step into the first line of the `numpy` module itself - which would be the expected behavior. If you keep stepping till you hit the eval, then a step-in on that would still give the same message about "source unavailable".

That part is correct, I think - it is a frame, it doesn't have a source that we can possibly fetch, but you can still e.g. inspect variables, and even step (blindly). It's not an internal debugger frame, and could very well be running user code, so hiding it would be confusing.

So I think the real bug here is that a `<string>` frame caused by eval/exec in stdlib code is not itself treated as stdlib, and the rest is by design. We can improve the error message though - if I remember correctly, the part before the colon is from VSC, and the part after is whatever error message we return from our `source` handler.

**Comment by int19h** [comment]
(Side note: in the future, we might want to try decompiling such frames with something like 'uncompyle6', and providing the synthesized source via source reference.)

**Comment by juliussimonelli** [comment]
When I try to add `"debugStdLib": true` to `launch.json` I get an error saying `Property debugStdLib is not allowed.` But based on [this](https://github.com/microsoft/vscode-python/issues/1322) issue it looks like should be working. I am using VSCode - Insiders Version 1.34.20-insider (1.34.20-insider).

**Comment by int19h** [comment]
We have changed it to `justMyCode` recently (which has to be set to `false` in this case), to be consistent with other VSC debuggers. You should get a prompt to fix it automatically in launch.json when you re-open your folder or workspace.

**Comment by juliussimonelli** [comment]
That worked. Thanks! By the way, when I click on "Quick Fix" it says `No code actions available.`

**Comment by parlad** [comment]
I was having the same issue 

`raw_input("enter input")` fixed this.

**Comment by bored002** [comment]
i'm trying the same thing but when debugging my python code with `"justMyCode": false` and launching the debug i'm getting a KeyError exception on `sys.path_importer_cache[path_item]` in pkgutil.py

**Comment by hellozhaihy** [comment]
In the latest vscode 1.36.1, adding `"justMyCode": false` to launch.json will sovle this problem.

**Comment by jajcayn** [comment]
Hey there,
same problem as @bored002: Getting some `Exception has occurred: KeyError` on `importer = sys.path_importer_cache[path_item]` in `pkgutil.py` when I add `"justMyCode": false` to the launch.json. Any advice on how to proceed?

Python 3.7.4 64bit
vscode 1.38.1
macOS 10.14.6

**Comment by int19h** [comment]
@jajcayn Just press F5 to continue running.

**Comment by W-void** [comment]
> We have changed it to `"justMyCode"` recently (which has to be set to `false` in this case), to be consistent with other VSC debuggers. You should get a prompt to fix it automatically in launch.json when you re-open your folder or workspace.

when I use np.clip(), there are some wrong :Could not load source '<__array_function__ internals>': Source unavailable. I add "justMyCode": false, but it's not useful

**Comment by Joy1112** [comment]
> > We have changed it to `"justMyCode"` recently (which has to be set to `false` in this case), to be consistent with other VSC debuggers. You should get a prompt to fix it automatically in launch.json when you re-open your folder or workspace.
> 
> when I use np.clip(), there are some wrong :Could not load source '<**array_function** internals>': Source unavailable. I add "justMyCode": false, but it's not useful

I also encountered the same problem and '"justMyCode": false' not working for me as well.

**Comment by Joy1112** [comment]
> > We have changed it to `"justMyCode"` recently (which has to be set to `false` in this case), to be consistent with other VSC debuggers. You should get a prompt to fix it automatically in launch.json when you re-open your folder or workspace.
> 
> when I use np.clip(), there are some wrong :Could not load source '<**array_function** internals>': Source unavailable. I add "justMyCode": false, but it's not useful

I also encountered the same problem and '"justMyCode": false' not working for me either.

**Comment by syhinex** [comment]
Same here, it was working fine with python 3.7. I updated Anaconda and python to 3.8 which broke things, dosn't matter what I set the value of "justMyCode", VScode refuses to step into the code and throws the 'Source unavailable' error.

**Comment by int19h** [comment]
To clarify: the error message itself relates only to the inability to fetch the source (because there isn't any for that frame). You have still successfully stepped in, and you can continue stepping in, or step out, or just continue running - it doesn't break the debug session as a whole.

**Comment by nandan-rpa** [comment]
> You are debugging non-user code, hence you need to turn that feature on.
> Add the following into `launch.json`
> 
> ```js
> "debugStdLib": true
> ```

It worked for me
I had code 'import pandas as pd'

but once on that step, we need to step-over else debugger gets into the 'pandas' routine..

**Comment by Sutyke** [comment]
Any idea about solution?

I'm getting similar error on ubuntu 18.04. When I step in to the code with VSCODE I get "Could not load source '<expr>': Source unavailable." . 

I would advice to use PyCharm as there there is excellent debugger where step in to the code works.

BTW: Is it possible to implement the same solution that is in pycharm, with 2 buttons for 1) step in and 2) step in my code?

**Comment by int19h** [comment]
We're showing the disassembled bytecode now, but the fundamental problem here is that we don't treat it as library code.
